ray-on-gke/user/kuberay-tpu-webhook/main.go

```diff
@@ -0,0 +1,242 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	ray "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// our representation of a pod slice
+// not necessarily true that worker group scheduled on 1 slice
+type Slice struct {
+	rayClusterName	string
+	groupName	string
+}
+
+var (
+	tpuResourceName = corev1.ResourceName("google.com/tpu")
+	certPath = "/etc/kuberay-tpu-webhook/tls/tls.crt"
+	keyPath = "/etc/kuberay-tpu-webhook/tls/tls.key"
+
+	// map of ray cluster names to # of workers created in the slice
+	sliceToWorkers map[Slice]int
+)
+
+// check if containers are requesting TPU resources
+func containerRequestingTPUs(containers ...corev1.Container) bool {
+	for _, container := range containers {
+		if l := container.Resources.Limits; l != nil {
+			if resource := l[tpuResourceName]; !resource.IsZero() {
+				return true
+			}
+		}
+		if r := container.Resources.Requests; r != nil {
+			if resource := r[tpuResourceName]; !resource.IsZero() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// unmarshal raycluster from admission request
+func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCluster, error) {
+	if admissionReview.Request.Kind.Kind != "RayCluster" {
+		return nil, fmt.Errorf("Expected RayCluster but got %s", admissionReview.Request.Kind.Kind)
+	}
+
+	rayCluster := ray.RayCluster{}
+	if err := json.Unmarshal(admissionReview.Request.Object.Raw, &rayCluster); err != nil {
+		return nil, err
+	}
+
+	return &rayCluster, nil
+}
+
+// add TPU_WORKER_HOSTNAMES to containers in a ray cluster
+func mutateRayCluster(
+	admissionReview *admissionv1.AdmissionReview,
+) (*admissionv1.AdmissionResponse, error) {
+	raycluster, _ := extractRayCluster(admissionReview)
+	patches := []map[string]interface{}{}
+
+	for i := 0; i < len(raycluster.Spec.WorkerGroupSpecs); i++ {
+		if(containerRequestingTPUs(raycluster.Spec.WorkerGroupSpecs[i].Template.Spec.Containers...)) {
+			numWorkers := int(*raycluster.Spec.WorkerGroupSpecs[i].Replicas)
+
+			// generate DNS hostnames
+			workerGroupName := raycluster.Spec.WorkerGroupSpecs[i].GroupName
+			serviceName := "headless-svc" // TODO: get headless-svc associated with workergroup
```

Comment:
Let's try to fix this before we submit.

Reply:
I believe this section of the code kind of relies on the changes we proposed to Kuberay. My understanding is that for multi-slice, we want to be creating a headless service with each workergroup and setting the name somewhere. Since we're not yet creating workergroups that way I just hardcoded the service name to the name of the one headless service we're creating.